### ISSUE TITLE
[Logging] Cannot use ConsoleAppender for Query and Security Logs

### ISSUE THREAD

**Issue by ikwattro** [neo4j/neo4j#13030]
**Version** : 5.3.0-enterprise
**Platform** : irrelevant

It seems the usage of `ConsoleAppender` is filtered out for anything else than `neo4j console` , is there a reason why ? The documentation is highlighting that all the Log4j appenders are supported and there is an example for the ConsoleAppender https://neo4j.com/docs/operations-manual/current/monitoring/logging/#_console_appender

The following `server-logs.xml` produces no logs for query and security logs

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!--

    Copyright (c) "Neo4j"
    Neo4j Sweden AB [http://neo4j.com]
    This file is a commercial add-on to Neo4j Enterprise Edition.

-->
<!--
    This is a log4j 2 configuration file.

    It is highly recommended to keep the original "debug.log" as is, to make sure enough data is captured in case
    of errors in a format that neo4j developers can work with.

    All configuration values can be queried with the lookup prefix "config:". You can for example, resolve
    the path to your neo4j home directory with ${config:dbms.directories.neo4j_home}.

    Please consult https://logging.apache.org/log4j/2.x/manual/configuration.html for instructions and
    available configuration options.
-->
<Configuration status="ERROR" monitorInterval="5" packages="org.neo4j.logging.log4j">
    <Appenders>
        <!-- Default debug.log, please keep -->
        <RollingRandomAccessFile name="DebugLog" fileName="${config:server.directories.logs}/debug.log"
                                 filePattern="$${config:server.directories.logs}/debug.log.%02i">
            <Neo4jDebugLogLayout pattern="%d{yyyy-MM-dd HH:mm:ss.SSSZ}{GMT+0} %-5p [%c{1.}] %m%n"/>
            <Policies>
                <SizeBasedTriggeringPolicy size="20 MB"/>
            </Policies>
            <DefaultRolloverStrategy fileIndex="min" max="7"/>
        </RollingRandomAccessFile>
        <Console name="ConsoleAppender" target="SYSTEM_OUT">
            <PatternLayout pattern="%d{yyyy-MM-dd HH:mm:ss.SSSZ}{GMT+0} %-5p %m%n"/>
        </Console>
    </Appenders>

    <Loggers>
        <!-- Log levels. One of DEBUG, INFO, WARN, ERROR or OFF -->

        <!-- The debug log is used as the root logger to catch everything -->
        <Root level="INFO">
            <AppenderRef ref="DebugLog"/> <!-- Keep this -->
        </Root>

        <!-- The query log, must be named "QueryLogger" -->
        <Logger name="QueryLogger" level="DEBUG" additivity="true">
            <AppenderRef ref="ConsoleAppender"/>
        </Logger>
        <!-- The security log, must be named "SecurityLogger" -->
        <Logger name="SecurityLogger" level="DEBUG" additivity="true">
            <AppenderRef ref="ConsoleAppender"/>
        </Logger>
    </Loggers>
</Configuration>
```

Reference to the code filtering the console appender https://github.com/neo4j/neo4j/blob/5.3/community/logging/src/main/java/org/neo4j/logging/log4j/LogConfig.java#L193

**Comment by klaren** [comment]
Hello and thanks for reporting!

There is more to it than that filtering, see https://github.com/neo4j/neo4j/blob/5.3/community/neo4j/src/main/java/org/neo4j/server/NeoBootstrapper.java#L145, so all console appenders would be redirected to the log files anyway. I believe the reason behind this was that there is no "natural" console attached to the `neo4j server`(daemonized), since that is highly platform and deployment dependent. In order to keep things as coherent as possible this is what we landed on.

That being said, the complete lack of warnings/log messages/documentation about this is really confusing, and at least that should be improved. I'll see to it!

Just out of curiosity, do you have a particular use case in mind when you want everything to be printed to stdout?

**Comment by ikwattro** [comment]
Hi @klaren Thanks for the clarification.

I definitely have use cases for stdout, mainly when Neo4j is deployed as a container. Having the ability to have logs produced to the container stdout means that it can leverage all the centralised logging infrastructure.

- We have Neo4j deployed on Kubernetes, we have nothing to do to send logs to loki for container logs
- We have Neo4j deployed on Azure App Services, having logs to Log Analytics is a one click option ( or 2 lines in a template ), while with files based, we have to create a new storage account, create a container and attach the container to the app service.

**Comment by klaren** [comment]
@ikwattro alright, that sounds reasonable. The neo4j docker images use `neo4j console` to run, so we have not encountered this limitation before in our testing setup. I'll see if there is a quick workaround until we can get this sorted.

**Comment by TomVanWemmel** [comment]
I'm also interested in a workaround. We are trying to do the same thing for HTTP logs.

**Comment by klaren** [comment]
@ikwattro @TomVanWemmel A well-tested solution would be to change your containers to use `neo4j console` akin to what neo4j docker and the init.d/systemd services use. Would this work for you?

If you need to emulate the `neo4j server` with console appenders you could spawn a process from the shell, either with `nohup` or the ampersand post-fix(`neo4j console &`) depending on the lifetime of the shell. A `neo4j console` process will react to `neo4j status` and `neo4j stop` the same way the server mode does, so the overall change would be minimal.

This only applies to Linux since Windows services do not have `stdout`/`stdin`/`stderr`. This is one of the reasons we opted for this behavior, to keep things coherent across platforms.

I'll update the documentation for the tarball installation with guidelines on when to use, or when not to use the different modes.

**Comment by ikwattro** [comment]
@klaren Not sure I fully understand this part : 

```
akin to what neo4j docker and the init.d/systemd services use. 
```

I'm already using the official neo4j docker today.

```
version: '3'
services:
  neo4j:
    image: neo4j:5.4.0-enterprise
    ports:
      - 7474:7474
      - 7473:7473
      - 6362:6362
      - 7687:7687
      - 7688:7688
    volumes:
      - "./backups:/backups"
      - "./seeds:/seeds"
      - "./server-logs.xml:/var/lib/neo4j/conf/server-logs.xml"
      - "./plugins:/plugins"
    environment:
      - NEO4J_dbms_databases_seed__from__uri__providers=URLConnectionSeedProvider,S3SeedProvider
      - NEO4J_AUTH=neo4j/password
      - NEO4J_ACCEPT_LICENSE_AGREEMENT=yes
 ```

**Comment by klaren** [comment]
@ikwattro The neo4j docker images should use `neo4j console`: https://github.com/neo4j/docker-neo4j/blob/master/docker-image-src/5/docker-entrypoint.sh#L624, so your console appenders should not be removed. Something else must be happening here. 

Could you run with the env `NEO4J_DEBUG=true` and verify that console is indeed used?

**Comment by ikwattro** [comment]
The logs show that it should start with `--console-mode` 

```
neo4j-5-logging-console-neo4j-1  | DEBUGGING ENABLED
neo4j-5-logging-console-neo4j-1  | Running as root user inside neo4j image
neo4j-5-logging-console-neo4j-1  | chowning /var/lib/neo4j recursively to neo4j:neo4j
neo4j-5-logging-console-neo4j-1  | Setting all files in /var/lib/neo4j/conf to permissions 600
neo4j-5-logging-console-neo4j-1  | checking /logs is writable
neo4j-5-logging-console-neo4j-1  | checking /data is writable
neo4j-5-logging-console-neo4j-1  | checking /data/databases is writable
neo4j-5-logging-console-neo4j-1  | checking /data/transactions is writable
neo4j-5-logging-console-neo4j-1  | Setting docker specific configuration overrides
neo4j-5-logging-console-neo4j-1  | Appended db.tx_log.rotation.retention_policy=100M size to /var/lib/neo4j/conf/neo4j.conf
neo4j-5-logging-console-neo4j-1  | Appended server.memory.pagecache.size=512M to /var/lib/neo4j/conf/neo4j.conf
neo4j-5-logging-console-neo4j-1  | Appended server.default_listen_address=0.0.0.0 to /var/lib/neo4j/conf/neo4j.conf
neo4j-5-logging-console-neo4j-1  | Setting docker specific Enterprise Edition overrides
neo4j-5-logging-console-neo4j-1  | Appended server.discovery.advertised_address=758d1fe1e89f:5000 to /var/lib/neo4j/conf/neo4j.conf
neo4j-5-logging-console-neo4j-1  | Appended server.cluster.advertised_address=758d1fe1e89f:6000 to /var/lib/neo4j/conf/neo4j.conf
neo4j-5-logging-console-neo4j-1  | Appended server.cluster.raft.advertised_address=758d1fe1e89f:7000 to /var/lib/neo4j/conf/neo4j.conf
neo4j-5-logging-console-neo4j-1  | Applying configuration settings that have been set using environment variables.
neo4j-5-logging-console-neo4j-1  | Removing existing setting for server.directories.logs in /var/lib/neo4j/conf/neo4j.conf
neo4j-5-logging-console-neo4j-1  | Appended server.directories.logs=/logs to /var/lib/neo4j/conf/neo4j.conf
neo4j-5-logging-console-neo4j-1  | NEO4J_AUTH has been parsed as user "neo4j", password "password", do_reset ""
neo4j-5-logging-console-neo4j-1  | Making sure /data/dbms is owned by neo4j:neo4j
neo4j-5-logging-console-neo4j-1  | Setting initial password
neo4j-5-logging-console-neo4j-1  | gosu neo4j:neo4j neo4j-admin dbms set-initial-password password --verbose
neo4j-5-logging-console-neo4j-1  | Executing command line: /opt/java/openjdk/bin/java -cp /var/lib/neo4j/plugins/*:/var/lib/neo4j/conf/*:/var/lib/neo4j/lib/* -XX:+UseParallelGC -XX:-OmitStackTraceInFastThrow -XX:+UnlockExperimentalVMOptions -XX:+TrustFinalNonStaticFields -XX:+DisableExplicitGC -Djdk.nio.maxCachedBufferSize=1024 -Dio.netty.tryReflectionSetAccessible=true -XX:+ExitOnOutOfMemoryError -Djdk.tls.ephemeralDHKeySize=2048 -XX:FlightRecorderOptions=stackdepth=256 -XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED -Dlog4j2.disable.jmx=true -Dfile.encoding=UTF-8 org.neo4j.cli.AdminTool dbms set-initial-password password --verbose
neo4j-5-logging-console-neo4j-1  | neo4j 5.4.0
neo4j-5-logging-console-neo4j-1  | VM Name: OpenJDK 64-Bit Server VM
neo4j-5-logging-console-neo4j-1  | VM Vendor: Eclipse Adoptium
neo4j-5-logging-console-neo4j-1  | VM Version: 17.0.5+8
neo4j-5-logging-console-neo4j-1  | JIT compiler: HotSpot 64-Bit Tiered Compilers
neo4j-5-logging-console-neo4j-1  | VM Arguments: [-XX:+UseParallelGC, -XX:-OmitStackTraceInFastThrow, -XX:+UnlockExperimentalVMOptions, -XX:+TrustFinalNonStaticFields, -XX:+DisableExplicitGC, -Djdk.nio.maxCachedBufferSize=1024, -Dio.netty.tryReflectionSetAccessible=true, -XX:+ExitOnOutOfMemoryError, -Djdk.tls.ephemeralDHKeySize=2048, -XX:FlightRecorderOptions=stackdepth=256, -XX:+UnlockDiagnosticVMOptions, -XX:+DebugNonSafepoints, --add-opens=java.base/java.nio=ALL-UNNAMED, --add-opens=java.base/java.io=ALL-UNNAMED, --add-opens=java.base/sun.nio.ch=ALL-UNNAMED, -Dlog4j2.disable.jmx=true, -Dfile.encoding=UTF-8]
neo4j-5-logging-console-neo4j-1  | Configuration files used (ordered by priority):
neo4j-5-logging-console-neo4j-1  | /var/lib/neo4j/conf/neo4j-admin.conf
neo4j-5-logging-console-neo4j-1  | /var/lib/neo4j/conf/neo4j.conf
neo4j-5-logging-console-neo4j-1  | --------------------
neo4j-5-logging-console-neo4j-1  | Changed password for user 'neo4j'. IMPORTANT: this change will only take effect if performed before the database is started for the first time.
neo4j-5-logging-console-neo4j-1  | getting full neo4j run command
neo4j-5-logging-console-neo4j-1  | exec gosu neo4j:neo4j /opt/java/openjdk/bin/java -cp /var/lib/neo4j/plugins/*:/var/lib/neo4j/conf/*:/var/lib/neo4j/lib/* -XX:+UseG1GC -XX:-OmitStackTraceInFastThrow -XX:+AlwaysPreTouch -XX:+UnlockExperimentalVMOptions -XX:+TrustFinalNonStaticFields -XX:+DisableExplicitGC -Djdk.nio.maxCachedBufferSize=1024 -Dio.netty.tryReflectionSetAccessible=true -Djdk.tls.ephemeralDHKeySize=2048 -Djdk.tls.rejectClientInitiatedRenegotiation=true -XX:FlightRecorderOptions=stackdepth=256 -XX:+UnlockDiagnosticVMOptions -XX:+DebugNonSafepoints --add-opens=java.base/java.nio=ALL-UNNAMED --add-opens=java.base/java.io=ALL-UNNAMED --add-opens=java.base/sun.nio.ch=ALL-UNNAMED -Dlog4j2.disable.jmx=true -Dfile.encoding=UTF-8 com.neo4j.server.enterprise.EnterpriseEntryPoint --home-dir=/var/lib/neo4j --config-dir=/var/lib/neo4j/conf --console-mode
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:33.523+0000 INFO  The license agreement was accepted with environment variable NEO4J_ACCEPT_LICENSE_AGREEMENT=yes when the Software was started.
neo4j-5-logging-console-neo4j-1  |
neo4j-5-logging-console-neo4j-1  |
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:33.534+0000 INFO  Starting...
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:35.333+0000 INFO  ======== Neo4j 5.4.0 ========
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:35.381+0000 INFO  This instance is ServerId{93c4a822} (93c4a822-ad0a-4ffb-9992-e958f7f8182e)
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:37.811+0000 INFO  Default database 'neo4j' is created
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:38.077+0000 INFO  Sending metrics to CSV file at /var/lib/neo4j/metrics
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:38.117+0000 INFO  Bolt enabled on 0.0.0.0:7687.
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:38.124+0000 INFO  Bolt (Routing) enabled on 0.0.0.0:7688.
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:38.745+0000 INFO  Remote interface available at http://localhost:7474/
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:38.748+0000 INFO  id: 05D227F66BC360848A9D7B4FC1BBCEC90185139F6F19F6D3953D97B25098547A
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:38.748+0000 INFO  name: system
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:38.749+0000 INFO  creationDate: 2023-02-03T12:55:36.851Z
neo4j-5-logging-console-neo4j-1  | 2023-02-03 12:55:38.749+0000 INFO  Started.
```

However, no logs are shown to the container logs.

Here is a sample compose project with the log4j2 xml config I'm using for testing

https://github.com/ikwattro/neo4j-docker-playground/tree/main/neo4j-5-logging-console

To be sure I also checked the server-logs.xml file is correctly mounted inside the container

```
$ docker exec -it neo4j-5-logging-console-neo4j-1 cat /var/lib/neo4j/conf/server-logs.xml
<?xml version="1.0" encoding="UTF-8"?>
<!--

    Copyright (c) "Neo4j"
    Neo4j Sweden AB [http://neo4j.com]
    This file is a commercial add-on to Neo4j Enterprise Edition.

-->
<!--
    This is a log4j 2 configuration file.

    It is highly recommended to keep the original "debug.log" as is, to make sure enough data is captured in case
    of errors in a format that neo4j developers can work with.

    All configuration values can be queried with the lookup prefix "config:". You can for example, resolve
    the path to your neo4j home directory with ${config:dbms.directories.neo4j_home}.

    Please consult https://logging.apache.org/log4j/2.x/manual/configuration.html for instructions and
    available configuration options.
-->
<Configuration status="ERROR" monitorInterval="5" packages="org.neo4j.logging.log4j">
    <Appenders>
        <!-- Default debug.log, please keep -->
        <RollingRandomAccessFile name="DebugLog" fileName="${config:server.directories.logs}/debug.log"
                                 filePattern="$${config:server.directories.logs}/debug.log.%02i">
            <Neo4jDebugLogLayout pattern="%d{yyyy-MM-dd HH:mm:ss.SSSZ}{GMT+0} %-5p [%c{1.}] %m%n"/>
            <Policies>
                <SizeBasedTriggeringPolicy size="20 MB"/>
            </Policies>
            <DefaultRolloverStrategy fileIndex="min" max="7"/>
        </RollingRandomAccessFile>
        <Console name="ConsoleAppender" target="SYSTEM_OUT">
            <PatternLayout pattern="%d{yyyy-MM-dd HH:mm:ss.SSSZ}{GMT+0} %-5p %m%n"/>
        </Console>
    </Appenders>

    <Loggers>
        <!-- Log levels. One of DEBUG, INFO, WARN, ERROR or OFF -->

        <!-- The debug log is used as the root logger to catch everything -->
        <Root level="INFO">
            <AppenderRef ref="DebugLog"/> <!-- Keep this -->
        </Root>

        <!-- The query log, must be named "QueryLogger" -->
        <Logger name="QueryLogger" level="DEBUG" additivity="true">
            <AppenderRef ref="ConsoleAppender"/>
        </Logger>
        <!-- The security log, must be named "SecurityLogger" -->
        <Logger name="SecurityLogger" level="DEBUG" additivity="true">
            <AppenderRef ref="ConsoleAppender"/>
        </Logger>
    </Loggers>
</Configuration>
```

**Comment by klaren** [comment]
@ikwattro Thanks for all the details, I have reproduced the behavior now. It appears that the `-console-mode` is not properly applied to the `server-logs.xml` and only affects the `user-logs.xml`. This is obviously a bug and not the intended behavior.

I'm working on a fix for it. Sadly I don't have a workaround for you that you can use in the meantime.

**Comment by ikwattro** [comment]
Thanks @klaren , no need for a workaround today, happy to wait for a fix when it comes.

Cheers !

**Comment by klaren** [comment]
@ikwattro Fix is part of the 5.6 release just released :tada:

**Comment by ikwattro** [comment]
Thanks @klaren , yes I'm waiting for it to be available on Docker hub to update my examples, thanks again !
